### PR TITLE
feat: preparatory work for realizing erc20

### PR DESCRIPTION
**Pull request by dreamer-zq**
Close：#389

### CODE REVIEW

File: proto/irismod/token/v1/event.proto [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+message EventDeployERC20 {
+    string symbol = 1;
+    string name = 2;
+    uint32 scale = 3;
+    string min_unit = 4;
+    string contract = 5;
+}
+
+// EventSwapToERC20 is an event emitted when swapping token from native token to ERC20.
+message EventSwapToERC20 {
```
Comment (scott-irisnet) [inline proto/irismod/token/v1/event.proto:21]: add a `string` attribute named `to_contract`

File: proto/irismod/token/v1/event.proto [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  cosmos.base.v1beta1.Coin amount = 1 [(gogoproto.nullable) = false];
+  string sender = 2 [ (cosmos_proto.scalar) = "cosmos.AddressString" ];
+  string receiver = 3;
+}
+
+
+// EventSwapFromERC20 is an event emitted when swapping token from ERC20 to native token.
+message EventSwapFromERC20 {
+  string contract = 1;
```
Comment (scott-irisnet) [inline proto/irismod/token/v1/event.proto:30]: rename `contract` to `from_contract`

File: proto/irismod/token/v1/event.proto [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  int64 amount = 2;
```
Comment (scott-irisnet) [inline proto/irismod/token/v1/event.proto:31]: change the type of `amount` to `cosmos.base.v1beta1.Coin`

File: proto/irismod/token/v1/tx.proto [hunk 4]
```diff
@@ -30,14 +30,24 @@ service Msg {
   rpc TransferTokenOwner(MsgTransferTokenOwner)
       returns (MsgTransferTokenOwnerResponse);
 
-  // SwapFeeToken defines a method for swap some fee token
+  // SwapFeeToken defines a method for swaping some fee token
```
Comment (scott-irisnet) [inline proto/irismod/token/v1/tx.proto:33]: .. defines a method for swapping between `IRIS` and `ERIS`

File: proto/irismod/token/v1/tx.proto [hunk 5]
```diff
@@ -30,14 +30,24 @@ service Msg {
...
   rpc SwapFeeToken(MsgSwapFeeToken) returns (MsgSwapFeeTokenResponse);
 
+  // SwapToERC20 defines a method for swaping some erc20 token
```
Comment (scott-irisnet) [inline proto/irismod/token/v1/tx.proto:36]: .. defines a method for swapping some native token to its ERC20 counterpart

File: proto/irismod/token/v1/tx.proto [hunk 6]
```diff
@@ -30,14 +30,24 @@ service Msg {
...
+  rpc SwapToERC20(MsgSwapToERC20) returns (MsgSwapToERC20Response);
+
+  // SwapFromERC20 defines a method for some native token
```
Comment (scott-irisnet) [inline proto/irismod/token/v1/tx.proto:39]: .. defines a method for swapping some ERC20 token to its native counterpart

File: proto/irismod/token/v1/tx.proto [hunk 7]
```diff
@@ -30,14 +30,24 @@ service Msg {
...
+  rpc SwapFromERC20(MsgSwapFromERC20) returns (MsgSwapFromERC20Response);
+
   // UpdateParams defines a governance operation for updating the token
   // module parameters. The authority is defined in the keeper.
   //
   // Since: cosmos-sdk 0.47
   rpc UpdateParams(MsgUpdateParams) returns (MsgUpdateParamsResponse);
+
+  // DeployERC20 defines a governance operation for deploying erc20 token for
+  // the native token
```
Comment (scott-irisnet) [inline proto/irismod/token/v1/tx.proto:49]: .. defines a governance operation for deploying an ERC20 contract that binds to a native token

File: proto/irismod/token/v1/tx.proto [hunk 8]
```diff
@@ -143,4 +153,41 @@ message MsgUpdateParams {
...
+}
+
+// MsgSwapToERC20Response defines the Msg/SwapToERC20 response type
+message MsgSwapToERC20Response {}
+
+// MsgSwapFromERC20 defines an SDK message for SwapFromERC20
+message MsgSwapFromERC20 {
+  option (cosmos.msg.v1.signer) = "sender";
+
+  string contract = 1;
```
Comment (scott-irisnet) [inline proto/irismod/token/v1/tx.proto:186]: remove `contract` as it is redundant

File: proto/irismod/token/v1/tx.proto [hunk 9]
```diff
@@ -143,4 +153,41 @@ message MsgUpdateParams {
...
+  int64 amount = 2;
```
Comment (scott-irisnet) [inline proto/irismod/token/v1/tx.proto:187]: change `int64 amount` to `Coin wanted_amount`

**Review by scott-irisnet — CHANGES_REQUESTED** [review]

File: contracts/Token.sol [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+     * - `to` cannot be the zero address.
+     * - `amount` caller must have a balance of at least `amount`.
+     * - `isERIS` whether to exchange for iris.
+     */
+    function swapToNative(
+        string memory to,
+        uint256 amount,
+        bool isERIS
+    ) public nonReentrant {
+        require(msg.sender.code.length == 0, "only called by EOA address");
```
Comment (mitch1024) [inline contracts/Token.sol:76]: What does `only called by EOA address` mean? Can't this be called by a 4337 contract?

File: contracts/Token.sol [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+     * Requirements:
+     *
+     * - `to` cannot be the zero address.
+     * - `amount` caller must have a balance of at least `amount`.
+     */
+    function swapToNative(
+        string memory to,
+        uint256 amount
+    ) public nonReentrant {
+        require(bytes(to).length > 0, "to must be vaild iaa address");
```
Comment (mitch1024) [inline contracts/Token.sol:74]:
```suggestion
        require(bytes(to).length > 0, "to must be a vaild iaa address");
```

**Review by scott-irisnet — APPROVED** [review]

**Pull request merged**